File: Java/Programmers/교점에별만들기.java
import java.util.*;

class Point {
    int x, y;
    Point(int x, int y) {
        this.x = x;
        this.y = y;
    }
}
class Solution {
    private static double[] findIntersection(int[] line1, int[] line2) {
        double a = line1[0], b = line1[1], e = line1[2];
        double c = line2[0], d = line2[1], f = line2[2];
        double x = (b * f - e * d) / (a * d - b * c);
        double y = (e * c - a * f) / (a * d - b * c);
        return new double[] {x, y};
    }
    
    public String[] solution(int[][] line) {
        List<Point> intersections = new ArrayList<>();
        for (int i = 0; i < line.length; i++) {
            for (int j = i + 1; j < line.length; j++) {
                double[] intersection = Solution.findIntersection(line[i], line[j]);
                double epsilon = 1e-9;
                if (Math.abs(intersection[0] - Math.round(intersection[0])) > epsilon || 
                    Math.abs(intersection[1] - Math.round(intersection[1])) > epsilon) {
                    continue;
                }

                int x = (int)Math.round(intersection[0]);
                int y = (int)Math.round(intersection[1]);
                intersections.add(new Point(x, y));
            }
        }
        int minX = intersections.stream().mapToInt(p -> p.x).min().orElse(0);
        int maxX = intersections.stream().mapToInt(p -> p.x).max().orElse(0);
        int minY = intersections.stream().mapToInt(p -> p.y).min().orElse(0);
        int maxY = intersections.stream().mapToInt(p -> p.y).max().orElse(0);
        char[][] graph = new char[maxY - minY + 1][maxX - minX + 1];
        for (int i = 0; i < graph.length; i++) {
            Arrays.fill(graph[i], '.');
        }
        for (int i = 0; i < intersections.size(); i++) {
            graph[intersections.get(i).y - minY][intersections.get(i).x - minX] = '*';
        }
        for (int i = 0; i < graph.length / 2; i++) {
            char[] temp = graph[i];
            graph[i] = graph[graph.length - i - 1];
            graph[graph.length - i - 1] = temp;
        }
        String[] answer = new String[graph.length];
        for (int i = 0; i < graph.length; i++) {
            answer[i] = new String(graph[i]);
        }
        return answer;
    }
}

File: swift/programmers/교점에 별 만들기.swift
// https://school.programmers.co.kr/learn/courses/30/lessons/87377

import Foundation

typealias Line = (a: Int, b: Int, c: Int)
typealias Point = (x: Int, y: Int)
func findIntersection(_ line1: [Int], _ line2: [Int]) -> (Double, Double) {
    let x = Double(line1[1] * line2[2] - line1[2] * line2[1]) / Double(line1[0] * line2[1] - line1[1] * line2[0])
    let y = Double(line1[2] * line2[0] - line1[0] * line2[2]) / Double(line1[0] * line2[1] - line1[1] * line2[0])
    return (x, y)
}
func solution(_ line:[[Int]]) -> [String] {
    var intersections = [Point]()
    for i in line.indices {
        for j in i + 1..<line.count {
            let intersection = findIntersection(line[i], line[j])
            guard intersection.0.truncatingRemainder(dividingBy: 1) == 0 && intersection.1.truncatingRemainder(dividingBy: 1) == 0 else { continue }
            intersections.append((Int(intersection.0), Int(intersection.1)))
        }
    }
    let maxX = intersections.max { $0.x < $1.x }?.x ?? 0
    let maxY = intersections.max { $0.y < $1.y }?.y ?? 0
    let minX = intersections.min { $0.x < $1.x }?.x ?? 0
    let minY = intersections.min { $0.y < $1.y }?.y ?? 0
    print(maxX, maxY, minX, minY)
    var graph = [[Character]](repeating: [Character](repeating: ".", count: maxX - minX + 1), count: maxY - minY + 1)
    // 1. minX만큼, minY만큼 뺀 만큼 shift 하기 (원점에 맞추기)
    // 2. 배열 순서 뒤집기
    // 3. x축은 column, y축은 row
    for point in intersections {
        graph[point.y - minY][point.x - minX] = "*"
    }
    let answer = graph.reversed().map { String($0) }
    return answer
}
